src/mbgl/text/bidi.hpp

```diff
@@ -14,17 +14,28 @@ class BiDiImpl;
 
 std::u16string applyArabicShaping(const std::u16string&);
 
+// StyledText pairs each code point in a string with an integer indicating
+// the styling options to use for rendering that code point
+// The data structure is intended to accomodate the reordering/interleaving
+// of formatting that can happen when BiDi rearranges inputs
+using StyledText = std::pair<std::u16string, std::vector<uint8_t>>;
```

Comment:
@ChrisLoer Since all symbol labels will be coerced to formatted text, are you concerned about memory consumption per-label increasing with each label being represented by a StyledText? The worst-case scenario is for a string to have a different style per-character, but I imagine that for the bulk of strings (those coerced) there will only be a single style. 
Would it be possible to maintain the styling options in a block format instead of per-code-point? the size of the vector would equal the number of sections and each entry would be the end index of the section with the corresponding index in the `sections` array. 

| Text              | hello | hola | こんにちは |
|-------------------|-------|------|------------|
| Section end index | 4     | 8    | 13?        |
| Section           | 0     | 1    | 2          |

`TaggedString` would need to perform a search linear in sections length for `getSectionAt()` but for the most common case that shouldn't be too bad.

Reply:
On the JS side I did a little experimentation with a data structure _kind of_ along the lines of what you're suggesting. It is a more memory-compact representation for the common case. What I was doing was iterating over the blocks as a whole, which made a lot of the iteration logic in the shaping code significantly more complicated. I think you're suggesting maintaining the same interface to `TaggedString` but just calculating the section index on-the-fly, so a CPU vs. memory trade off. That seems reasonable, but I don't know which would win without actually doing the experiment.

FWIW, on the JS side I looked at overall memory consumption in a full-screen map before and after this change, and couldn't detect any difference. It should be 1 byte-per-character, and I'd guess a full screen map has maybe on the order of 10k characters worth of text?